pkg/function/restore_data_using_kopia_server.go

```diff
@@ -0,0 +1,261 @@
+		vols,
+		podOverride,
+	)
+}
+
+func restoreDataFromServer(
+	ctx context.Context,
+	cli kubernetes.Interface,
+	hostname,
+	image,
+	jobPrefix,
+	namespace,
+	restorePath,
+	serverAddress,
+	fingerprint,
+	snapID,
+	username,
+	userPassphrase string,
+	sparseRestore bool,
+	vols map[string]string,
+	podOverride crv1alpha1.JSONMap,
+) (map[string]any, error) {
+	for pvc := range vols {
+		if _, err := cli.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvc, metav1.GetOptions{}); err != nil {
+			return nil, errors.Wrap(err, "Failed to retrieve PVC from namespace: "+namespace+" name: "+pvc)
+		}
+	}
+
+	options := &kube.PodOptions{
+		Namespace:    namespace,
+		GenerateName: jobPrefix,
+		Image:        image,
+		Command:      []string{"bash", "-c", "tail -f /dev/null"},
+		Volumes:      vols,
+		PodOverride:  podOverride,
+	}
+
+	pr := kube.NewPodRunner(cli, options)
+	podFunc := restoreDataFromServerPodFunc(
+		cli,
+		hostname,
+		namespace,
+		restorePath,
+		serverAddress,
+		fingerprint,
+		snapID,
+		username,
+		userPassphrase,
+		sparseRestore,
+	)
+	return pr.Run(ctx, podFunc)
+}
+
+func restoreDataFromServerPodFunc(
+	cli kubernetes.Interface,
+	hostname,
+	namespace,
+	restorePath,
+	serverAddress,
+	fingerprint,
+	snapID,
+	username,
+	userPassphrase string,
+	sparseRestore bool,
+) func(ctx context.Context, pod *corev1.Pod) (map[string]any, error) {
+	return func(ctx context.Context, pod *corev1.Pod) (map[string]any, error) {
+		if err := kube.WaitForPodReady(ctx, cli, pod.Namespace, pod.Name); err != nil {
+			return nil, errors.Wrap(err, "Failed while waiting for Pod: "+pod.Name+" to be ready")
+		}
+
+		contentCacheMB, metadataCacheMB := kopiacmd.GetCacheSizeSettingsForSnapshot()
+		configFile, logDirectory := kankopia.CustomConfigFileAndLogDirectory(hostname)
+
+		cmd := kopiacmd.RepositoryConnectServerCommand(
+			kopiacmd.RepositoryServerCommandArgs{
+				UserPassword:    userPassphrase,
+				ConfigFilePath:  configFile,
+				LogDirectory:    logDirectory,
+				CacheDirectory:  kopiacmd.DefaultCacheDirectory,
+				Hostname:        hostname,
+				ServerURL:       serverAddress,
+				Fingerprint:     fingerprint,
+				Username:        username,
+				ContentCacheMB:  contentCacheMB,
+				MetadataCacheMB: metadataCacheMB,
+			})
+		stdout, stderr, err := kube.Exec(cli, namespace, pod.Name, pod.Spec.Containers[0].Name, cmd, nil)
+		format.Log(pod.Name, pod.Spec.Containers[0].Name, stdout)
+		format.Log(pod.Name, pod.Spec.Containers[0].Name, stderr)
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to connect to Kopia API server")
+		}
+
+		cmd = kopiacmd.SnapshotRestore(
+			kopiacmd.SnapshotRestoreCommandArgs{
+				CommandArgs: &kopiacmd.CommandArgs{
+					RepoPassword:   "",
+					ConfigFilePath: configFile,
+					LogDirectory:   logDirectory,
+				},
+				SnapID:                 snapID,
+				TargetPath:             restorePath,
+				SparseRestore:          sparseRestore,
+				IgnorePermissionErrors: true,
```

Comment:
Ideally, we should not ignore permission errors

Reply:
I think there are some problems upstream. Keeping it true as of now !